### PR TITLE
Feature/issue 61

### PR DESCRIPTION
**Pull request by dhivesh**
Dimmer samples projects:
- Spring boot project
- Standalone project
Tests the multiple examples and combinations with Dimmer.

In progress: README.md of each sample
Add one more example with the same feature name and 2 different operations for the spring-boot project.

### CODE REVIEW

File: dimmer-samples/dimmer-spring-boot-configuration/src/main/resources/application-dev.properties [hunk 1]
```diff
@@ -0,0 +1 @@
+endpointAvailableForIntegration=
```
Comment (dieppa) [inline dimmer-samples/dimmer-spring-boot-configuration/src/main/resources/application-dev.properties]: if empty, why don't remove it?

File: dimmer-samples/dimmer-spring-boot-configuration/src/main/resources/application.properties [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# H2
+spring.h2.console.enabled=true
+spring.h2.console.path=/h2
+# Datasource
+spring.datasource.url=jdbc:h2:mem:prod
+spring.datasource.driverClassName=org.h2.Driver
+spring.datasource.username=sa
+spring.datasource.password=
+spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
+endpointAvailableForIntegration=
```
Comment (dieppa) [inline dimmer-samples/dimmer-spring-boot-configuration/src/main/resources/application.properties]: if empty, why don't remove it?

File: dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/DimmerApplicationIntegrationTestDefault.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import static com.github.cloudyrock.dimmer.samples.configuration.DimmerConfiguration.DEV;
+import static com.github.cloudyrock.dimmer.samples.controller.UserController.USERS_PATH;
+import static org.hamcrest.core.Is.is;
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.junit.Assert.assertThat;
+
+@EnableAutoConfiguration
+@ActiveProfiles(DEFAULT)
+public class DimmerApplicationIntegrationTestDefault extends DimmerApplicationIntegrationTest {
+
```
Comment (dieppa) [inline dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/DimmerApplicationIntegrationTestDefault.java]: What is that pattern TestDefault, testDev, etc.???
I would rather use something like:
DimmerApplicationDefaultEnvironmentITest.

Using ITest por integration tests and UTest for unit tests

File: dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/DimmerApplicationIntegrationTestDev.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package com.github.cloudyrock.dimmer.samples;
+
+import org.json.JSONObject;
+import org.junit.Test;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.http.*;
+import org.springframework.test.context.ActiveProfiles;
```
Comment (dieppa) [inline dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/DimmerApplicationIntegrationTestDev.java]: wildcard import and optimise all the imports

Comment (dieppa) [reply dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/DimmerApplicationIntegrationTestDev.java]: There are more wildcard in other tests/classes, remove them ;)

File: dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/controller/UserControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+@RunWith(MockitoJUnitRunner.class)
+public class UserControllerTest {
+
+    public static final String USER1 = "DIMMER";
+
+    //This line helps initialising a default mocked environment for Dimmer annotations.
+    @Mock
+    private FeatureExecutor featureExecutor = DimmerBuilder.local().defaultEnvironment().buildWithDefaultEnvironment();
+
```
Comment (dieppa) [inline dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/controller/UserControllerTest.java]: If you mock with annotation, with assign another value?

File: dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/DimmerApplicationIntegrationTestDefault.java (hunk 3 above)
Comment (dieppa) [reply dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/DimmerApplicationIntegrationTestDefault.java]: optimise imports

File: dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/DimmerApplicationIntegrationTestDefault.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.boot.web.server.LocalServerPort;
+import org.springframework.http.*;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.web.client.RestClientException;
+
```
Comment (dieppa) [inline dimmer-samples/dimmer-spring-boot-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/DimmerApplicationIntegrationTestDefault.java]: optimise imports

File: dimmer-samples/dimmer-standalone-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/StandaloneConfigurationExampleTest.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    StandaloneConfigurationExample standaloneConfigurationExample;
+
+    @Mock
+    private PrintStream printStreamMock;
+
+    @Before
+    public void setUp() {
+        initMocks(this);
+        System.setOut(printStreamMock);
+    }
```
Comment (dieppa) [inline dimmer-samples/dimmer-standalone-configuration/src/test/java/com/github/cloudyrock/dimmer/samples/StandaloneConfigurationExampleTest.java]: 👍

File: dimmer-samples/dimmer-spring-boot-configuration/pom.xml [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        <dependency>
+            <groupId>com.github.cloudyrock.dimmer</groupId>
+            <artifactId>dimmer</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-jersey</artifactId>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (dhivesh) [reply dimmer-samples/dimmer-spring-boot-configuration/pom.xml]: True, done

File: dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/configuration/DimmerConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    @Bean
+    @Profile(DEV)
+    public FeatureExecutor dimmerBuilderDefault(Environment environment) {
+        LOGGER.info("Starting Dimmer Builder on environment {}", environment.getActiveProfiles()[0]);
+        return DimmerBuilder.local()
+                .environments(DEV)
+                    .featureWithException(GET_USERS, GET_USERS_CONTROLLER, NotImplementedException.class)
+                    .featureWithValue(ADD_USER, CREATE_USER_CONTROLLER, new UserApiResponse(1L, "MOCKED VALUE"))
+                .build(DEV);
+    }
```
[the remark this answers is not in the document]
Comment (dhivesh) [reply dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/configuration/DimmerConfiguration.java]: Done

File: dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/configuration/DimmerConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Bean
+    @Profile(PROD)
+    public FeatureExecutor dimmerBuilderProd(Environment environment) {
+        LOGGER.info("Starting Dimmer Builder on environment {}", environment.getActiveProfiles()[0]);
+        return DimmerBuilder.local()
+                .environments(PROD)
+                .featureWithDefaultExceptionConditional(isIntegrationAvailable(), ADD_USER, CREATE_USER_CONTROLLER)
+                .build(PROD);
+    }
```
[the remark this answers is not in the document]
Comment (dhivesh) [reply dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/configuration/DimmerConfiguration.java]: Done

File: dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/configuration/DimmerConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    // Conditional example of a behaviour: configure the feature behaviour depending on the value of a parameter in configuration
+    // Some other examples where this could be applied would be trying to reach out a service or business logic.
+    private boolean isIntegrationAvailable() {
+        LOGGER.info("Endpoint {} is available", endpointAvailableForIntegration);
+        if (endpointAvailableForIntegration != null && !endpointAvailableForIntegration.isEmpty()) {
+            return true;
+        } else {
+            return false;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (dhivesh) [reply dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/configuration/DimmerConfiguration.java]: Done

File: dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/configuration/NotImplementedException.java [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package com.github.cloudyrock.dimmer.samples.configuration;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value = HttpStatus.NOT_IMPLEMENTED)
+public class NotImplementedException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (dhivesh) [reply dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/configuration/NotImplementedException.java]: Yes. In fact the body of the response contains the messages:
```json
{"timestamp":"2018-09-16T15:44:46.795+0000","status":501,"error":"Not Implemented","message":"Feature is work in progress :)","path":"/users"}
```

File: dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/repository/User.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+@Table(name = "User")
+public class User {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long id;
+
+    @Size
+    private String name;
+
```
[the remark this answers is not in the document]
Comment (dhivesh) [reply dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/repository/User.java]: Agreed, but not necessary for this project  :)

File: dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/repository/UserRepository.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface UserRepository extends JpaRepository<User, Long> {
+
+    User findByName(String name);
+}
```
[the remark this answers is not in the document]
Comment (dhivesh) [reply dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/repository/UserRepository.java]: For future enhancements :)

[the remark this answers is not in the document]
Comment (dhivesh) [reply dimmer-samples/dimmer-spring-boot-configuration/src/main/java/com/github/cloudyrock/dimmer/samples/repository/UserRepository.java]: Done :)

**Comment by dhivesh** [comment]
The samples have been moved to a different repository: https://github.com/cloudyrock/dimmer-project-samples

**Pull request closed without merging**